FLINK-18851 [runtime] Add checkpoint type to checkpoint history entries in Web UI

gm7y8 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Add checkpoint type to checkpoint history entries in Web UI

## Brief change log

Add checkpoint type to checkpoint history entries in Web UI


## Verifying this change

*(Please pick either of the following options)*

This change is already covered by existing tests, such as *(please describe tests)*.
Also, we can deploy Flink job with savepoint or checkpoint enable, then you should be able to see checkpoint Type in the flink history tab.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: don't know
  - The runtime per-record code paths (performance sensitive):  no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper:  no
  - The S3 file system connector:  no 

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? not applicable

gm7y8 commented:
@flinkbot run azure

XComp reviewed (changes requested):
Thanks @gm7y8 for picking up this feature. I went through the code and added some remarks.
Additionally, keep in mind that the [REST API documentation](https://github.com/apache/flink/tree/master/flink-docs#rest-api-documentation) needs to be regenerated as well when updating the API.

`mvn package -Dgenerate-rest-docs -pl flink-docs -am -nsu -DskipTests` will generate `docs/_includes/generated/rest_v1_dispatcher.html`. The change needs to committed.

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -113,6 +114,9 @@
 	private final int numAckSubtasks;
 
+	@JsonProperty(FIELD_NAME_CHECK_POINT_TYPE)
+	private final String checkPointType;
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -83,6 +82,8 @@
 
 	public static final String FIELD_NAME_TASKS = "tasks";
 
+	public static final String FIELD_NAME_CHECK_POINT_TYPE = "check_point_type";
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -18,6 +18,14 @@
 
 package org.apache.flink.runtime.rest.messages.checkpoints;
 
+import java.util.Collection;
```
done

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -18,15 +18,15 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import org.apache.flink.runtime.checkpoint.CheckpointStatsStatus;
```
The imports are still not compliant with the CheckStyle format defined for Flink. Please use [Flink's CheckStyle format definition](https://ci.apache.org/projects/flink/flink-docs-stable/flinkDev/ide_setup.html#checkstyle-for-java) to fix the formatting.

gm7y8 replied:
done

XComp reviewed (changes requested):
Thanks for your changes, @gm7y8 . A few remarks:
* The CheckStyle format is still not applied.
* It would be good to apply the `checkPointType` -> `checkpointType` changes in all of your newly added code lines.
* `docs/_includes/generated/rest_v1_dispatcher.html` still needs to be generated using `mvn package -Dgenerate-rest-docs -pl flink-docs -am -nsu -DskipTests`.

gm7y8 commented:
@flinkbot run azure

XComp reviewed (changes requested):
Thanks for looking into it @gm7y8 . I added a few more comments. Please remember to regenerate `docs/_includes/generated/rest_v1_dispatcher.html` and `flink-runtime-web/src/test/resources/rest_api_v1.snapshot` after you're done applying all changes.

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -18,32 +18,22 @@
-import org.apache.flink.runtime.checkpoint.FailedCheckpointStats;
-import org.apache.flink.runtime.checkpoint.PendingCheckpointStats;
-import org.apache.flink.runtime.checkpoint.TaskStateStats;
+import org.apache.flink.runtime.checkpoint.*;
```
done

gm7y8 commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -82,6 +82,7 @@ protected CheckpointingStatistics getTestResponseInstance() throws Exception {
 			0L,
 			10,
 			10,
+			 "Checkpoint",
```
done

gm7y8 commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -97,6 +98,7 @@ protected CheckpointingStatistics getTestResponseInstance() throws Exception {
 			0L,
 			9,
 			9,
+			"Savepoint",
```
done

gm7y8 commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -112,6 +114,7 @@ protected CheckpointingStatistics getTestResponseInstance() throws Exception {
 			0L,
 			11,
 			9,
+			"Checkpoint",
```
done

gm7y8 commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -133,6 +136,7 @@ protected CheckpointingStatistics getTestResponseInstance() throws Exception {
 			0L,
 			10,
 			10,
+			"Checkpoint",
```
done

gm7y8 commented on flink-runtime-web/src/test/resources/rest_api_v1.snapshot:
```diff
@@ -1254,6 +1263,9 @@
         "num_acknowledged_subtasks" : {
           "type" : "integer"
         },
+        "check_point_type" : {
```
done.. should these steps be documented in the developer guide https://ci.apache.org/projects/flink/flink-docs-master/flinkDev/building.html .. let me if I can add them here?

azagrebin replied:
It should be probably documented in `flink-runtime-web/README.md` if not yet anywhere else
what do you think @zentol ?
@gm7y8 if it is not documented anywhere else, you can create a JIRA issue for that and then open PR.

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -113,23 +116,27 @@
 	private final int numAckSubtasks;
 
+	@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE)
+	private final CheckpointType checkPointType;
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -113,23 +116,27 @@
+		@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
+		@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+		@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkPointType,
+		@JsonDeserialize(keyUsing = JobVertexIDKeyDeserializer.class) @JsonProperty(FIELD_NAME_TASKS) Map<JobVertexID, TaskCheckpointStatistics> checkpointStatisticsPerTask) {
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -113,23 +116,27 @@
+		@JsonProperty(FIELD_NAME_ALIGNMENT_BUFFERED) long alignmentBuffered,
+		@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
+		@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+		@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkPointType,
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -140,6 +147,7 @@ private CheckpointStatistics(
 		this.alignmentBuffered = alignmentBuffered;
 		this.numSubtasks = numSubtasks;
 		this.numAckSubtasks = numAckSubtasks;
+		this.checkPointType = checkPointType;
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -179,6 +187,10 @@ public int getNumAckSubtasks() {
 	}
 
+	public CheckpointType getCheckPointType() {
+		return checkPointType;
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -203,12 +215,13 @@ public boolean equals(Object o) {
 			numSubtasks == that.numSubtasks &&
 			numAckSubtasks == that.numAckSubtasks &&
 			status == that.status &&
+			Objects.equals(checkPointType, that.checkPointType) &&
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -203,12 +215,13 @@ public boolean equals(Object o) {
 	@Override
 	public int hashCode() {
-		return Objects.hash(id, status, savepoint, triggerTimestamp, latestAckTimestamp, stateSize, duration, alignmentBuffered, numSubtasks, numAckSubtasks, checkpointStatisticsPerTask);
+		return Objects.hash(id, status, savepoint, triggerTimestamp, latestAckTimestamp, stateSize, duration, alignmentBuffered, numSubtasks, numAckSubtasks, checkPointType, checkpointStatisticsPerTask);
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -330,6 +346,7 @@ public CompletedCheckpointStatistics(
 			@JsonProperty(FIELD_NAME_ALIGNMENT_BUFFERED) long alignmentBuffered,
 			@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
 			@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+			@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkPointType,
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -409,6 +427,7 @@ public FailedCheckpointStatistics(
 			@JsonProperty(FIELD_NAME_ALIGNMENT_BUFFERED) long alignmentBuffered,
 			@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
 			@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+			@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkPointType,
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -423,6 +442,7 @@ public FailedCheckpointStatistics(
 				alignmentBuffered,
 				numSubtasks,
 				numAckSubtasks,
+				checkPointType,
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -477,6 +497,7 @@ public PendingCheckpointStatistics(
 			@JsonProperty(FIELD_NAME_ALIGNMENT_BUFFERED) long alignmentBuffered,
 			@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
 			@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+			@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkPointType,
```
done

gm7y8 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -489,6 +510,7 @@ public PendingCheckpointStatistics(
 				alignmentBuffered,
 				numSubtasks,
 				numAckSubtasks,
+				checkPointType,
```
done

gm7y8 commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointingStatisticsTest.java:
```diff
@@ -27,6 +28,8 @@
 import java.util.Map;
 
+
+
```
yes I believe, removed these empty lines

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -113,23 +116,27 @@
-			@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
-			@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
-			@JsonDeserialize(keyUsing = JobVertexIDKeyDeserializer.class) @JsonProperty(FIELD_NAME_TASKS) Map<JobVertexID, TaskCheckpointStatistics> checkpointStatisticsPerTask) {
+		@JsonProperty(FIELD_NAME_ID) long id,
+		@JsonProperty(FIELD_NAME_STATUS) CheckpointStatsStatus status,
+		@JsonProperty(FIELD_NAME_IS_SAVEPOINT) boolean savepoint,
+		@JsonProperty(FIELD_NAME_TRIGGER_TIMESTAMP) long triggerTimestamp,
+		@JsonProperty(FIELD_NAME_LATEST_ACK_TIMESTAMP) long latestAckTimestamp,
+		@JsonProperty(FIELD_NAME_STATE_SIZE) long stateSize,
+		@JsonProperty(FIELD_NAME_DURATION) long duration,
+		@JsonProperty(FIELD_NAME_ALIGNMENT_BUFFERED) long alignmentBuffered,
+		@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
+		@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+		@JsonProperty(FIELD_NAME_CHECKPOINT_TYPE) CheckpointType checkpointType,
+		@JsonDeserialize(keyUsing = JobVertexIDKeyDeserializer.class) @JsonProperty(FIELD_NAME_TASKS) Map<JobVertexID, TaskCheckpointStatistics> checkpointStatisticsPerTask) {
```
Please fix the formatting here.

gm7y8 replied:
done

XComp reviewed (changes requested):
Thanks @gm7y8 for your changes. There are only two minor formatting things left which need to be addressed. Additionally, please update the commit message to comply to [Flink's commit format](https://flink.apache.org/contributing/contribute-documentation.html#submit-your-contribution). It should look like `[FLINK-18851][runtime-web] ...`.

I'm gonna ask @vthinkxie to review the `web-dashboard` changes in the meantime. For me, they look good.

gm7y8 commented on flink-runtime-web/src/test/resources/rest_api_v1.snapshot:
```diff
@@ -3082,4 +3098,4 @@
     }
   } ]
-}
+}
```
done

vthinkxie commented:
> Thanks @gm7y8 for your changes. There are only two minor formatting things left which need to be addressed. Additionally, please update the commit message to comply to [Flink's commit format](https://flink.apache.org/contributing/contribute-documentation.html#submit-your-contribution). It should look like `[FLINK-18851][runtime-web] ...`.
> 
> I'm gonna ask @vthinkxie to review the `web-dashboard` changes in the meantime. For me, they look good.

The frontend code looks good to me

gm7y8 commented:
@XComp I have taken care of commit message and other formatting issues.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -229,8 +242,8 @@ public static CheckpointStatistics generateCheckpointStatistics(AbstractCheckpoi
 					new TaskCheckpointStatistics(
-							checkpointStats.getCheckpointId(),
-							checkpointStats.getStatus(),
+						checkpointStats.getCheckpointId(),
+						checkpointStats.getStatus(),
```
this change also looks unrelated.
it looks like a minor code style fix which we usually do not do to preserve git history.

if we do the code style fixes then it is usually either
- massively in files which have a lot of code style violation in separate commit
- or close to related changes

gm7y8 replied:
Thanks for pointing out. let me fix this.

gm7y8 replied:
reverted back.

azagrebin reviewed (changes requested):
Thanks for this improvement @gm7y8
Overall, it looks good to me.

@XComp @vthinkxie Did someone check that the change works in UI?

I left comments about changes which are probably unrelated.
After addressing these comments/question, I can merge the PR.

Also, not sure what the first 'merge' commit means, I can remove it before merging.

gm7y8 commented on flink-runtime-web/src/test/resources/rest_api_v1.snapshot:
```diff
@@ -557,7 +557,7 @@
               "status" : {
                 "type" : "string",
-                "enum" : [ "CREATED", "RUNNING", "FAILING", "FAILED", "CANCELLING", "CANCELED", "FINISHED", "RESTARTING", "SUSPENDED", "RECONCILING" ]
+                "enum" : [ "INITIALIZING", "CREATED", "RUNNING", "FAILING", "FAILED", "CANCELLING", "CANCELED", "FINISHED", "RESTARTING", "SUSPENDED", "RECONCILING" ]
```
this was auto-generated code.. not related to this change.. will open jira issue and commit this change separately.

XComp replied:
This was already fixed in PR #13403 and merged into `master`. Rebasing the branch should fix this diff.

gm7y8 replied:
@XComp PR #13403  only fixed html documentation.. here it's missing in rest_api_v1.snapshot

XComp commented:
> > Thanks @gm7y8 for your changes. There are only two minor formatting things left which need to be addressed. Additionally, please update the commit message to comply to [Flink's commit format](https://flink.apache.org/contributing/contribute-documentation.html#submit-your-contribution). It should look like `[FLINK-18851][runtime-web] ...`.
> > I'm gonna ask @vthinkxie to review the `web-dashboard` changes in the meantime. For me, they look good.
> 
> The frontend code looks good to me

Reconsidering the change: @vthinkxie what about exposing this information to the Checkpoint history page as well?

XComp commented:
> Thanks for this improvement @gm7y8
> Overall, it looks good to me.
> 
> @XComp @vthinkxie Did someone check that the change works in UI?
> 
> I left comments about changes which are probably unrelated.
> After addressing these comments/question, I can merge the PR.
> 
> Also, not sure what the first 'merge' commit means, I can remove it before merging.

I ran the Web UI including the changes. It looks like expected.

gm7y8 commented:
@XComp @azagrebin created to fix bug FLINK-19501 to fix the missing enum.. other PR just addresses only the html part not the snapshot document

gm7y8 commented:
> Thanks for this improvement @gm7y8
> Overall, it looks good to me.
> 
> @XComp @vthinkxie Did someone check that the change works in UI?
> 
> I left comments about changes which are probably unrelated.
> After addressing these comments/question, I can merge the PR.
> 
> Also, not sure what the first 'merge' commit means, I can remove it before merging.

First Merge commit is created when I synced fork with the main repo.

AHeise commented:
I just checked the ticket and unfortunately the original intent was missed: it's not apparent to the user if a checkpoint is aligned or unaligned.

gm7y8 commented:
@AHeise sorry for the delay to respond back. it should simple fix in the UI layer. I would take the opinion of @XComp @vthinkxie if it ok with the above change.

XComp commented:
Hi @gm7y8 , please go ahead with implementing the request @AHeise brought up considering the text change @AHeise proposed:

> Possible values should be “aligned checkpoint”, “unaligned checkpoint”, “savepoint”, "sync_savepoint" (which I'd rename to "savepoint on cancel" to make it less technical and more user-friendly in the UI).

XComp commented:
@gm7y8 What's the status? Would it be possible for you to finish this feature within this week? Otherwise, I would take over by Wednesday to finish it prior to the feature freeze for `1.12` probably happening next week.

gm7y8 commented:
@XComp I was able to identify the code fix. I working to unit test it had some issue with Flink set up to start a job with checkpoint and savepoint. it might take a day or so as I am doing it for the first time.

XComp commented:
> @XComp I was able to identify the code fix. I working to unit test it had some issue with Flink set up to start a job with checkpoint and savepoint. it might take a day or so as I am doing it for the first time.

Hi @gm7y8 , thanks for getting back to us. Let's see if we can finish this before the codefreeze next week. Here a few remarks:
* The build is failing for me locally (`mvn -pl flink-runtime-web -Dfast -DskipTests install`) with the following error message:
```
[ERROR] ERROR in src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkPointConfig' does not exist on type 'JobCheckpointsDetailComponent'.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
[ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
```
  It's always good run a final `mvn install` at least on the modules you touched to verify your changes. Alternatively (or rather as an additional tool), you can setup Azure CI on your fork. That will run a full test covering all modules for each commit on your fork. Check out the [tutorial for setting up Azure CI](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).
* The changes like [job-checkpoints-detail.component.html](https://github.com/apache/flink/pull/13458/commits/0b8e369aa5c8c4e4a8c370b4b848c0f197fcdb9d#diff-a1c50fd814fe8a1c03567f9c71877873251223cfaf32c7067068f7f30d8c5c42R25) does not reflect @AHeise's request as you use uppercased tokens here. @AHeise suggested normal casing (i.e. lowercase) as it helps to improve the readability.
* Additionally, I realized that you didn't address all points of the PR description template. May you fix that?

Thanks.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -154,6 +161,7 @@ private CheckpointStatistics(
 		this.persistedData = persistedData;
 		this.numSubtasks = numSubtasks;
 		this.numAckSubtasks = numAckSubtasks;
+		this.checkpointType = checkpointType;
```
I forgot to mention in the other review: It's always good to call `Preconditions.checkNotNull` on nullable parameters. See other `status` as an example.

gm7y8 replied:
sure I can do that.

XComp reviewed (changes requested):
Thanks @gm7y8 . The changes look good. I just added a few minor things. Could you please update the PR description in addition to that answering all the open questions?

gm7y8 commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-checkpoint.ts:
```diff
@@ -114,6 +115,8 @@ export interface CheckPointConfigInterface {
     enabled: boolean;
     delete_on_cancellation: boolean;
   };
+  state_backend: string;
```
actually, this field is missing from the rest API object spec and this interface so just add it keep them in sync

gm7y8 commented:
> > @XComp I was able to identify the code fix. I working to unit test it had some issue with Flink set up to start a job with checkpoint and savepoint. it might take a day or so as I am doing it for the first time.
> 
> Hi @gm7y8 , thanks for getting back to us. Let's see if we can finish this before the codefreeze next week. Here a few remarks:
> 
> * The build is failing for me locally (`mvn -pl flink-runtime-web -Dfast -DskipTests install`) with the following error message:
> 
> ```
> [ERROR] ERROR in src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkPointConfig' does not exist on type 'JobCheckpointsDetailComponent'.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> ```
> 
> It's always good run a final `mvn install` at least on the modules you touched to verify your changes. Alternatively (or rather as an additional tool), you can setup Azure CI on your fork. That will run a full test covering all modules for each commit on your fork. Check out the [tutorial for setting up Azure CI](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).
> 
> * The changes like [job-checkpoints-detail.component.html](https://github.com/apache/flink/pull/13458/commits/0b8e369aa5c8c4e4a8c370b4b848c0f197fcdb9d#diff-a1c50fd814fe8a1c03567f9c71877873251223cfaf32c7067068f7f30d8c5c42R25) does not reflect @AHeise's request as you use uppercased tokens here. @AHeise suggested normal casing (i.e. lowercase) as it helps to improve the readability.
> * Additionally, I realized that you didn't address all points of the PR description template. May you fix that?
> 
> Thanks.

Sure, I fix the build issue u mentioned, and the build has passed now. Also will change the letter to lowercase.

gm7y8 commented:
@flinkbot approve description

gm7y8 commented:
> > > @XComp I was able to identify the code fix. I working to unit test it had some issue with Flink set up to start a job with checkpoint and savepoint. it might take a day or so as I am doing it for the first time.
> > 
> > 
> > Hi @gm7y8 , thanks for getting back to us. Let's see if we can finish this before the codefreeze next week. Here a few remarks:
> > 
> > * The build is failing for me locally (`mvn -pl flink-runtime-web -Dfast -DskipTests install`) with the following error message:
> > 
> > ```
> > [ERROR] ERROR in src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkPointConfig' does not exist on type 'JobCheckpointsDetailComponent'.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Element implicitly has an 'any' type because type 'CheckPointDetailInterface' has no index signature.
> > [ERROR] src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html(25,38): : Property 'checkpoint_type' does not exist on type 'JobCheckpointsDetailComponent'.
> > ```
> > 
> > 
> > It's always good run a final `mvn install` at least on the modules you touched to verify your changes. Alternatively (or rather as an additional tool), you can setup Azure CI on your fork. That will run a full test covering all modules for each commit on your fork. Check out the [tutorial for setting up Azure CI](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).
> > 
> > * The changes like [job-checkpoints-detail.component.html](https://github.com/apache/flink/pull/13458/commits/0b8e369aa5c8c4e4a8c370b4b848c0f197fcdb9d#diff-a1c50fd814fe8a1c03567f9c71877873251223cfaf32c7067068f7f30d8c5c42R25) does not reflect @AHeise's request as you use uppercased tokens here. @AHeise suggested normal casing (i.e. lowercase) as it helps to improve the readability.
> > * Additionally, I realized that you didn't address all points of the PR description template. May you fix that?
> > 
> > Thanks.
> 
> Sure, I fixed the build issue u mentioned, and the build has passed now. Also will change the letter to lowercase.

Also updated the description to reflect PR guidelines..

XComp commented:
Thanks, @gm7y8 . It looks good from my side. @vthinkxie: Could you have a brief look over the frontend code diff? I guess there have been some changes after your last review.

gm7y8 commented:
@XComp  You are welcome! I have also verified fix with Word Count job

![Uploading Screen Shot 2020-10-22 at 12.23.01 AM.png…]()

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.html:
```diff
@@ -21,6 +21,8 @@
     <nz-divider nzType="vertical"></nz-divider>
     <strong>Discarded:</strong> {{checkPointDetail?.discarded || '-'}}
+    <nz-divider nzType="vertical"></nz-divider>
+    <strong>Checkpoint Type:</strong> {{checkPointDetail?.checkpoint_type === "CHECKPOINT" ? (checkPointConfig?.unaligned_checkpoints ? 'unaligned checkpoint' : 'aligned checkpoint') : (checkPointDetail?.checkpoint_type === "SYNC_SAVEPOINT" ? 'savepoint on cancel' : 'savepoint')|| '-'}}
```
Hi @gm7y8 
it would be better to move this calculation into typescript file as a getter

gm7y8 replied:
makes sense. I just moved this into function in the ts file.

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.ts:
```diff
@@ -80,5 +83,25 @@ export class JobCheckpointsDetailComponent implements OnInit {
+    });
+  }
+
+  checkPointType(){
```
it would be better to checkPointType as a value other than a function here
then move the calculation to line 65

gm7y8 replied:
done.

gm7y8 commented:
@flinkbot run azure

gm7y8 commented:
@XComp For some reason its failing in the HBase repo i retriggered the build

gm7y8 commented:
@XComp @vthinkxie I have addressed the requested changes. Build also has passed.

XComp commented:
Thanks @gm7y8 . Additionally, I verified manually that the changes result in the expected behavior. Let's wait for @vthinkxie to get back to us to review the code itself.

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.ts:
```diff
@@ -57,9 +60,23 @@ export class JobCheckpointsDetailComponent implements OnInit {
   refresh() {
```
Hi @gm7y8 
could you try to replace the refresh function with the codes below

```ts

  refresh() {
    this.isLoading = true;
    if (this.jobDetail && this.jobDetail.jid) {
      forkJoin([
        this.jobService.loadCheckpointConfig(this.jobDetail.jid),
        this.jobService.loadCheckpointDetails(this.jobDetail.jid, this.checkPoint.id)
      ]).subscribe(
        ([config, detail]) => {
          this.checkPointConfig = config;
          this.checkPointDetail = detail;
          if (this.checkPointDetail.checkpoint_type === 'CHECKPOINT') {
            if (this.checkPointConfig.unaligned_checkpoints) {
              this.checkPointType = 'unaligned checkpoint';
            } else {
              this.checkPointType = 'aligned checkpoint';
            }
          } else if (this.checkPointDetail.checkpoint_type === 'SYNC_SAVEPOINT') {
            this.checkPointType = 'savepoint on cancel';
          } else if (this.checkPointDetail.checkpoint_type === 'SAVEPOINT') {
            this.checkPointType = 'savepoint';
          } else {
            this.checkPointType = '-';
          }
          this.isLoading = false;
          this.cdr.markForCheck();
        },
        () => {
          this.isLoading = false;
          this.cdr.markForCheck();
        }
      );
    }
  }

```

gm7y8 replied:
done!

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.ts:
```diff
@@ -80,5 +97,9 @@ export class JobCheckpointsDetailComponent implements OnInit {
       this.cdr.markForCheck();
       this.refresh();
     });
+    this.jobService.loadCheckpointConfig(this.jobDetail.jid).subscribe(config => {
```
these new lines could be removed after refactoring refresh func

gm7y8 replied:
done

gm7y8 commented:
@flinkbot run azure

gm7y8 commented:
> Thanks @gm7y8 . Additionally, I verified manually that the changes result in the expected behavior. Let's wait for @vthinkxie to get back to us to review the code itself.

@XComp @vthinkxie  I just verified the fix after making suggested enhancements. it is working as expected.

vthinkxie commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/detail/job-checkpoints-detail.component.ts:
```diff
@@ -55,21 +59,38 @@ export class JobCheckpointsDetailComponent implements OnInit {
+          }
+        );
+      }
     }
```
There are two extra spaces in each line, the others look good to me

gm7y8 replied:
@vthinkxie I have fixed the spaces.. my IntelliJ is not enforcing the check style, I assume it is due to the community edition doesn't support ts code auto format.

vthinkxie replied:
Hi, I have a submit a commit here
https://github.com/vthinkxie/flink/commit/0a513eb9ac724b2a15cf4c95acdf22e5a9456d10
could you have a look and apply this patch, there is still some format error

gm7y8 replied:
sure!

gm7y8 commented:
@flinkbot run azure

rmetzger commented:
Thanks a lot for the contribution and the reviews so far. I assume this PR is ready to merge?

gm7y8 commented:
@rmetzger I believe we should be good to merge. Thanks a lot @XComp, @vthinkxie for your time and patience.

rmetzger commented:
I checked the code and tried it out. Looks good. Thanks a lot. 

Merging ...
